scale-decode/src/visitor/types/variant.rs

```diff
@@ -14,43 +14,42 @@
 // limitations under the License.
 
 use crate::visitor::{Composite, DecodeError};
-use crate::Field;
-use scale_info::form::PortableForm;
-use scale_info::{Path, PortableRegistry, TypeDefVariant};
+use scale_type_resolver::{FieldIter, TypeResolver, VariantIter};
 
 /// A representation of the a variant type.
-pub struct Variant<'scale, 'info> {
+pub struct Variant<'scale, 'resolver, R: TypeResolver> {
     bytes: &'scale [u8],
-    variant: &'info scale_info::Variant<PortableForm>,
-    fields: Composite<'scale, 'info>,
+    variant_name: &'resolver str,
+    variant_index: u8,
+    fields: Composite<'scale, 'resolver, R>,
 }
 
-impl<'scale, 'info> Variant<'scale, 'info> {
-    pub(crate) fn new(
+impl<'scale, 'resolver, R: TypeResolver> Variant<'scale, 'resolver, R> {
+    pub(crate) fn new<
+        Fields: FieldIter<'resolver, R::TypeId> + 'resolver,
+        Variants: VariantIter<'resolver, Fields>,
+    >(
         bytes: &'scale [u8],
-        path: &'info Path<PortableForm>,
-        ty: &'info TypeDefVariant<PortableForm>,
-        types: &'info PortableRegistry,
-    ) -> Result<Variant<'scale, 'info>, DecodeError> {
+        mut variants: Variants,
+        types: &'resolver R,
+    ) -> Result<Variant<'scale, 'resolver, R>, DecodeError> {
         let index = *bytes.first().ok_or(DecodeError::NotEnoughInput)?;
         let item_bytes = &bytes[1..];
 
         // Does a variant exist with the index we're looking for?
-        let variant = ty
-            .variants
-            .iter()
+        let mut variant = variants
             .find(|v| v.index == index)
-            .ok_or_else(|| DecodeError::VariantNotFound(index, ty.clone()))?;
+            .ok_or_else(|| DecodeError::VariantNotFound(index))?;
 
         // Allow decoding of the fields:
-        let mut fields_iter = variant.fields.iter().map(|f| Field::new(f.ty.id, f.name.as_deref()));
-        let fields = Composite::new(item_bytes, path, &mut fields_iter, types, false);
+        // let mut fields_iter = variant.fields.map(|f| Field::new(f.id, f.name.as_deref()));
```

Comment:
Should this commented out line still be removed?

Reply:
Ooh yup, it's not needed any more :)